### ISSUE TITLE
Inbox Discovery/Delivery doesn't fully jive with LDN

### ISSUE THREAD

**Issue by gobengo** [w3c-social/activitypump#162]
From https://w3c.github.io/activitypub/#delivery

> NOTE: Relationship to Linked Data Notifications
> While it is not required reading to understand this specification, it is worth noting that ActivityPub's targeting and delivery mechanism overlaps with the Linked Data Notifications specification, and the two specifications may interoperably combined. In particular, the inbox property is the same between ActivityPub and Linked Data Notifications, and the targeting and delivery systems described in this document are supported by Linked Data Notifications.

Right now ActivityPub only says this about inbox discovery (given one URL, determining the URL of the inbox to which targeted Activities should be delivered):

> The inbox is discovered through the inbox property of an actor's profile.

which I believe is sort of only describing how to determine the inbox if that resource comes back as JSON.

But LDN discovery has this step as well "make an HTTP GET request on the target URL and, for RDF sources [RDF 1.1], the Inbox is the object of the predicate http://www.w3.org/ns/ldp#inbox."

By specifying the RDF predicate and not just the 'property', there's a way of discovering the inbox resource for other Content-Types for which the "actor's profile" might be serialized (like text/html + RDFa, turtle, or perfectly-valid ways of using JSON-LD in which 'http://www.w3.org/ns/ldp#inbox' is not mapped to the JSON 'inbox' key, but instead to something like 'ldp:inbox'.

Are you in favor of adding a similar step to inbox discovery that supports discovery-via-RDF predicate? Or maybe just wholesale delegating to LDN's discovery algorithm (which would also include HTTP Link Headers)?

This would, for example, allow for interop where you could post an activity to an ActivityPub outbox that targets one of @csarven's https://dokie.li/ documents, and ActivityPub delivery would deliver it to the LDN inboxes he points to via HTML+RDFa.

**Comment by rhiaro** [comment]
You're right, that totally slipped through. We just slimmed down the [AS2 support stuff in LDN](https://www.w3.org/TR/ldn/#activity-streams-2-0-support) to move the details to Social Web Protocols, but I haven't completely finished writing it [there](https://w3c-social.github.io/social-web-protocols/#delivery) yet.

> This would, for example, allow for interop where you could post an activity to an ActivityPub outbox that targets one of @csarven's https://dokie.li/ documents, and ActivityPub delivery would deliver it to the LDN inboxes he points to via HTML+RDFa.

This is _exactly_ what I want from life. I don't think we can require ActivityPub implementations to support all RDF parsing though, only JSON. Though this is on the server side so maybe it's not a huge stretch (as opposed to clients, which are more difficult). But right now the compromise is that LDN publishers who want to guarantee that naive AP (ie. plain JSON ish) implementations will discover their inbox MUST publish it using _compact_ JSON-LD (then looking for just `inbox` as a property will work), and that AP notification-sending implementations SHOULD look for `inbox` as an RDF property, and if they can't and fall back to plain JSON they might miss some. It's not ideal.

Perhaps what AP needs to add is that last SHOULD.

But otherwise, this issue is for SWP because I need to expand and reorganise this plain-JSON/JSON-LD interop stuff - would you mind raising it there?

**Comment by rhiaro** [comment]
Oh oh wait, the reason this wasn't such a huge deal was because in AP the targets are _always_ user profiles. If you're publishing a user profile with an inbox, you probably know it's an AP one, so you're going to publish it with compact JSON-LD anyway. AP servers doing discover shouldn't then hit up on anything unexpected. I think?

**Comment by rhiaro** [comment]
Okay so actually it's a client problem after all. Just to run through this.

> Clients are responsible for addressing new Activites appropriately. - [Client to Server interactions](https://w3c.github.io/activitypub/#client-to-server-interactions)

A new object is created which is `inReplyTo` a dokieli article. The client SHOULD fetch this and look for an `actor` or `attributedTo` property, and add this to the `to` or `cc` of the object. If it's a naive plain JSON client (likely, as many pump.io ones are desktop implementations in eg. C++) and the dokieli article is published in HTML+RDFa, even if it uses AS2, it won't be able to address it properly. So the target won't even hit the server, so the server won't have a chance to attempt to send an LDN.

_However_ if the dokieli article is connegged and can return its metadata as compact JSON-LD, the client will find the `attributedTo` and add it to the `to` field.

The creating server outbox will then see this, do discovery on _the user profile URL_ and find the inbox there.

**Comment by rhiaro** [comment]
What this boils down to I think is that LDN says "targeting is at the sender's discretion" and AP senders' "discretion" is to decide the target will always be ~~a user profile~~ found in the `to`, `cc` or `bcc` fields.

I think AP has it covered with:

> To some extent, this is dependent upon the particular client implementation, but clients must be aware that the server will only forward new Activities to addressees in the to, cc and bcc fields. 

Clients that know AS2 is also JSON-LD can look harder for targets, basically.

**Comment by gobengo** [comment]
tl;dr - I propose that AP specifies that inbox discovery SHOULD be done using LDN inbox discovery if one cannot be determined via JSON .inbox. This would make them jive more better.

@rhiaro 

> But right now the compromise is that LDN publishers who want to guarantee that naive AP (ie. plain JSON ish) implementations will discover their inbox MUST publish it using compact JSON-LD 

~~It doesn't say MUST be compact in AP spec~~ (While this is true, it's moot because as @rhiaro points out below, AS2 documents must be JSON-LD compacted with a normative context). But even if it did, I believe JSON-LD compaction is a two-argument operation: the graph document as well as a "context" to describe _how_ to compact it. i.e. there are infinite ways of compacting a JSON-LD document such that the JSON key for the 'https://www.w3.org/ns/activitypub#inbox' relation is not 'inbox'. See this example in the [Playground](http://json-ld.org/playground/#startTab=tab-compacted&json-ld=%7B%22%40context%22%3A%5B%22https%3A%2F%2Fwww.w3.org%2Fns%2Factivitystreams%22%2C%7B%22activitypub%22%3A%22https%3A%2F%2Fwww.w3.org%2Fns%2Factivitypub%23%22%2C%22inbox%22%3A%22activitypub%3Ainbox%22%2C%22outbox%22%3A%22activitypub%3Aoutbox%22%7D%5D%2C%22type%22%3A%22Service%22%2C%22name%22%3A%22distbin%22%2C%22summary%22%3A%22A%20public%20service%20to%20store%20and%20retrieve%20posts%20and%20enable%20%28federated%2C%20standards-compliant%29%20social%20interaction%20around%20them%22%2C%22inbox%22%3A%22%2Factivitypub%2Finbox%22%2C%22outbox%22%3A%22%2Factivitypub%2Foutbox%22%2C%22recent%22%3A%22%2Frecent%22%7D&frame=%7B%7D&context=%7B%22activitypub%22%3A%22https%3A%2F%2Fwww.w3.org%2Fns%2Factivitypub%23%22%2C%22lol-not-inbox%22%3A%22activitypub%3Ainbox%22%2C%22outbox%22%3A%22activitypub%3Aoutbox%22%7D).

Though ActivityPub could require that the graph be compacted with a particular, specified context like

``` json
{
  "inbox": "https://www.w3.org/ns/activitypub#inbox",
  "outbox": "https://www.w3.org/ns/activitypub#outbox"
}
```

> and that AP notification-sending implementations SHOULD look for inbox as an RDF property

Yeah, it would be good to add that SHOULD. On what it would say, though, there's another issue. Right now the URIs for AP 'inbox' and LDN 'inbox' are different. So which should RDF-aware AP clients look for, and in what order?

I would be in favor of wholly delegating to LDN#inbox. i.e. removing activitypub#inbox URI, or in any semantics marking it sameAs LDN#inbox. And in the AP spec, say "If the discovery client is not RDF-aware, request as application/json and look for .inbox. If it is, do exactly what LDN discovery says over here)"

This would have the added benefit, which I find pretty important, of supporting discovery for resources whose owners can only add metadata via HTML (and not ask their engineers to please support content negotiation to JSON-LD...) such as most articles on the web that are rendered via a CMS.

> But otherwise, this issue is for SWP because I need to expand and reorganise this plain-JSON/JSON-LD interop stuff - would you mind raising it there?

I don't really know what to file there, sorry <3. I guess because the proposal immediately above tries to just fix this ambiguity by having AP wholly delegate RDF-aware discovery to LDN, instead of using SWP to clarify.

> What this boils down to I think is that LDN says "targeting is at the sender's discretion" and AP senders' "discretion" is to decide the target will always

I disagree. It seems to me there are three discrete steps: Targeting (which you describe), _discovery_ of inbox URL (which is where my issue lies), and delivery. I'm proposing AP have a richer _discovery_ algorithm (or delegate to LDN's), not targeting.

**Comment by rhiaro** [comment]
> It doesn't say MUST be compact in AP spec

All AP JSON is AS2, which does say must be compact, and thus compacted with the AS2 `@context`.

That activitypub context isn't actually right, AP terms are going into the AS2 JSON-LD context. I think there's another issue to fix that. And in the AS2 context, `as:inbox` is (or will be when I add it) aliased to `ldp:inbox`. So for RDF processors they are the same. Sorry, you're missing some pieces of the puzzle here, it's helpful that you're flagging them.

So what's going on is that non-RDF-aware clients are requesting `activity+json` which is a compact JSON-LD document with the AS2 `@context` which those clients would (safely) ignore.

> This would have the added benefit, which I find pretty important, 

I agree, which is why we support this in LDN. AP (and AS2) takes the pure JSON-based approach though and expects JSON first, other representations second. That's why it's relevant that AP targeting is only done on user profiles, because AP user profiles are _always_ available as JSON, and the discovery follows from that.

**Comment by rhiaro** [comment]
Oh, I agree that a note to say "if you're looking for a JSON `inbox` property and you don't find it, maybe think about RDF" could be good.

**Comment by cwebber** [comment]
> Oh, I agree that a note to say "if you're looking for a JSON inbox property and you don't find it, maybe think about RDF" could be good.

That'll be nice to add, and I'll improve the note about the relationship to LDN to say that.  However, it won't be that ActivityPub will put targeting via RDF as a SHOULD, because many ActivityPub implementations won't be RDF-aware.  That's by design.  If they are, great!  But it's not an expectation, and I believe SHOULD sets an expectation.

**Comment by cwebber** [comment]
I pushed an update to the spec, as you can see above.  What do you think?

**Comment by gobengo** [comment]
Hmm. Good conversation.

**@cwebber wrt your commit**

I think that the following to clauses are inconsistent, but your commit makes it better than before.
"it is worth noting that ActivityPub's targeting and delivery mechanism overlaps with the Linked Data Notifications specification, and the two specifications may interoperably combined"
"Linked Data Notifications also supports a number of RDF serializations which are not required for ActivityPub implementations."

It's not really interoperability if you can only interoperate with 1/3rd (probably less) of resources pointing to LDN inboxes.

(See next section because I think we'd calculate that ratio differently.)

---

**on goals for how many things can be AP-delivered to**

I think Amy alludes to something important here that highlights a difference in our goals. Upon sending an activity to my outbox, I want it to notify any resource linked to that activity that advertises an LDN inbox. The current scope of ActivityPub only cares about notifying "ActivityPub user profile" resources that advertises an LDN inbox, which we say must know how to respond to requests with a pretty obscure Accept header of "application/ld+json; profile="https://www.w3.org/ns/activitystreams#"

> That's why it's relevant that AP targeting is only done on user profiles, because AP user profiles are always available as JSON, and the discovery follows from that.

I wish it were not the case that "AP targeting is only done on user profiles", and encourage a broader scope of thinking, and I don't think it costs anything. I'm trying to implement toward a federated demo. There are currently **zero** ActivityPub "user profiles" on the web to demo with (ok fine, [one](http://distbin.com/)). Why constrain the domain to just "user profiles". There are dozens of other non-user-profile 'types' of resources live on the web right now that are ready and able to receive LDNs, and advertise an inbox. I think that this thinking about only wanting to deliver to "user profiles" is sort of narrowly thinking just about MediaGoblin and rhiaro.co.uk use cases. Why shouldn't ActivityPub be useful for delivering to the inbox of https://dokie.li/ even though it's not a "user profile"?

---

**on "pure JSON"**

> All AP JSON is AS2, which does say must be compact, and thus compacted with the AS2 @context.

Didn't remember that. Thanks.

> AP (and AS2) takes the pure JSON-based approach

With regard to AP, this isn't really true. The bridge to JSON-LD has already been crossed. In several places (3.2, 7), it's specified that Activitypub clients **MUST** POST/GET things as `application/ld+json; profile="https://www.w3.org/ns/activitystreams#"`

Furthermore, there's not a single mention of "application/json" in the whole spec. Why do we care about "pure JSON"? Is it so that developers don't get confused or scared off when they see "ld+json"? If so, we have work to do to become more pure. But I think that while this ideal of pure JSON is important for adoption of the AS2 vocabulary, it's much less important in an networked application protocol that involves, in-part, what is almost like an extension of AS2 (outbox/followers/[etc](https://w3c.github.io/activitypub/#actor-objects)), and AS2 says to do extensions via JSON-LD.

---

**on where activitypub should define it's names**
tldr - in activitypub's namespace, not someone else's. or don't define names that aren't needed

> That activitypub context isn't actually right, AP terms are going into the AS2 JSON-LD context. I think there's another issue to fix that. And in the AS2 context, as:inbox is (or will be when I add it) aliased to ldp:inbox. So for RDF processors they are the same. Sorry, you're missing some pieces of the puzzle here, it's helpful that you're flagging them.

It's odd to me that activitypub would create terms in the AS2 context. Well, maybe not, but it begs some questions.
- So we're clear, Is this what you intend to do? https://github.com/w3c/activitystreams/compare/master...gobengo:patch-1
- If you claim as2#inbox in the JSON-LD context, do you also plan to define 'inbox' in the human-readable spec? If yes, will it be defined to be something LDN-related or more general? If not, what happens if AS2 eventually wants to use the term 'inbox' for something more generic?
- This would be the first term in the AS2 context that is defined elsewhere (LDN spec). And LDN is only a WD whereas AS2 is a quite stable CR. Up until this point AS2 has been remarkably dependency-free (other than some light mentions of JSON-LD, which is a recommendation). I worry that making AS2 depend on LDN puts AS2 at risk.

After thinking a little bit, I'm feeling like the constraint of "All AP JSON is AS2" sort of paints things into the corner without much benefit.

It seems to me that a cleaner route would be for activitypub to publish it's own context which can be used alongside as2. Something like this:

https://www.w3.org/ns/activitypub

``` json
{
  "@context": {
      "inbox": "http://www.w3.org/ns/ldp#inbox"
  }
}
```

Require _ActivityPub_ clients to request with 'Accept: application/ld+json; profile="https://www.w3.org/ns/activitypub'.
Require AP servers responding to that Accept to perform JSON-LD compaction with 

``` json
{
  "@context": ["https://www.w3.org/ns/activitystreams", "https://www.w3.org/ns/activitypub"]
}
```

(requests coming in with Accept: application/json should be required to be compacted the same way)

Yes, this invalidates "All AP JSON is AS2" (barely). But if I'm discovering an inbox for purposes of ActivityPub delivery, I don't care if it's AS2. I care that it's useful for ActivityPub. And, by the way, every other key except 'inbox' is exactly the same (even without JSON-LD awareness by the consumer) as AS2. And if the consumer is JSON-LD aware, they can tell which terms are as2 and which are activitypub anyway. All this without requiring AS2 to have its context defined across several specs and dependent on LDN.

There is perhaps an even more simple solution to avoid AP defining terms in AS2's ns, allow for AS2 to define a different "inbox" term in the future, maintain "All AP JSON is AS2", preserve a trivial discovery algorithm for JSON-LD-unaware clients, and maintain compatibility with [the JSON-LD part of] LDN's discovery algorithm. And that is to keep things exactly as they are in AP today, but instead of saying

> The inbox is discovered through the inbox property of an actor's profile.

say

> The inbox is discovered through the "http://www.w3.org/ns/ldp#inbox" property of an actor's profile.

No need to touch AS2's context just for the aesthetic benefit of a JSON object having valid string properties that happen to also be URLs.

I am not sure which of those two solutions is best. But the latter one seems simplest for now. I can empathize with really really really wanting to avoid URL keys, in which case I think the former is a reasonable way of accommodating all concerns.

**Comment by cwebber** [comment]
> With regard to AP, this isn't really true. The bridge to JSON-LD has already been crossed. In several places (3.2, 7), it's specified that Activitypub clients MUST POST/GET things as `application/ld+json; profile="https://www.w3.org/ns/activitystreams#"`

ActivityPub is very intently built so that implementers who don't understand RDF can work with it.  That's not going to change.  We switched from using the activitystreams specific mimetype to using the `application/ld+json` with a profile because it would allow other linked data systems to work with AP at theoretically no cost.  But if the cost turns out to be "well, now that you're partway here, that means you must accept an entire world of RDF!" then it turns out, that's too much of a cost.

The document might not say "we take a pure JSON" approach, but the spec is designed with that in mind.  We're specifically developing to support "linke-data-naive" implementations.  Pubstrate, which doesn't yet have a public instance, but is in this boat, is in this camp.  MediaGoblin probably will be once it goes live.  Neither project is set up to traverse a world of triples type approach.  And it would take tremendous rearchitecting to make either do it.

But, the nice thing about json-ld is that linked data systems can at least understand how to read our stuff and perceive our data that way.  And we can take advantage of nice linked data properties like namespaced, unambiguous terms and etc.

ActivityPub isn't going to become a "full world of RDF" system.  But we currently have the olive branch so that people who have bought into that can cooperate with us, if they are willing to treat us as a subset.

I worry if you take the "full RDF or nothing!" approach, you'll end up with "or nothing".  We have a nice venn diagram of interop here.  I'd hate to kill that.

**Comment by cwebber** [comment]
I also think that json-ld _is_ linked data... I hope I'm not wrong there.  What I mean to say here is, in many applications I won't have the tooling to buy into a much broader "RDF as a series of triples / graph" set of tooling.  ActivityPub is set up explicitly so that people who only have SQL or document store type systems can use the existing schemas they have, and tooling that's familiar.  The promise of ActivityPub is: you can achieve federation while just at that world view.  I'm totally open to encouraging people to explore _beyond_ that world view, but if we drop this implicit assumption that you can do ActivityPub with those above assumptions, then we've dropped a lot of what ActivityPub delivers, and I think we'll lose many implementors.

**Comment by akuckartz** [comment]
JSON-LD is _both_ RDF _and_ pure JSON. Therefore implementations should be possible with which both communities can live with. That at least is the intention.

**Comment by cwebber** [comment]
@akuckartz Right!

So I think in sum, ActivityPub allows for pure JSON implementations that don't know about RDF, but for implementations that _are_ RDF-aware, at least they can read ActivityPub's content (even if not all ActivityPub implementations can consume non-json-ld representations from other implementations).

But by _pointing to_ LDN, maybe some implementations will explore other formats, too.

**Comment by cwebber** [comment]
So I guess I'd ask at this point, is there any action to be taken on this ticket?  It seems the primary argument seems to be "encourage users to consume many types of RDF", and I've added encouragement (though I haven't made it SHOULD/MUST).  Is that good enough, or is there more to be done?

**Comment by gobengo** [comment]
tl;dr ya, @rhiaro should respond to the thing I bolded below.
Other than that you can close and I won't be offended.

---

I filed this issue originally to point out that AP discovery won't always find a validly advertised LDN inbox. In my subjective opinion, that make this note, which remains in the spec, misleading:

> it is worth noting that ActivityPub's targeting and delivery mechanism overlaps with the Linked Data Notifications specification, and the two specifications may interoperably combined

I still think that given the state of the normative parts of AP, the above clause is misleading. I guess if I had to propose what would make it more clear, I'd change "overlaps with" to "is a subset of". And strike the second sentence, because this basically means most AP impls won't interoperate with most resources with LDN documents. I think the note in the spec should call out that AP outboxes are _not_ required to interoperate with LDN instead of implying that things will be all hunky-dory.

At the same time, I think I've been heard. I mostly filed this to get some clarity on what to expect out of AP and what my implementation should do. I now have that clarity. Anything else on that issue is fair to consider editorial.

---

The "on where activitypub should define it's names" bit is also kind of ranty on my part, and has been ignored so far. That's fine. It's also sort of an editor's decision.

However, I would still like to understand what's going to happen. This part above was meant to be addressed to @rhiaro :

> - So we're clear, Is this what you intend to do? https://github.com/w3c/activitystreams/compare/master...gobengo:patch-1

**I do sort of want to leave the issue open until I get an answer there.** The first section here was about how an AP implementation (that doesn't grok RDF) would try to find an LDN inbox (and most of the time it wont). This second bit is about whether an RDF-aware LDN client would be able to discover an LDN inbox for all valid AP 'user profile's. The current state of things is that it couldn't, but Amy implies she wants it to be able to. If that **ACTION** is tracked somewhere else, then this ticket can close.

---

I will drop what I mean in "on goals for how many things can be AP-delivered to" for now. I think it's not relevant to the topic of this issue.

**Comment by cwebber** [comment]
> The "on where activitypub should define it's names" bit is also kind of ranty on my part, and has been ignored so far. That's fine. It's also sort of an editor's decision.

It's been discussed elsewhere though, and we had a face to face resolution on it.  #131 was an issue filed/resolved as an output of that face to face discussion.  The group agreed to this, including discussion with ActivityStreams co-editors @jasnell and @evanp that we would add these terms to the namespace as an extension.

So as for where ActivityPub will define "inbox", what has been discussed is that the ActivityStreams json-ld document will alias "inbox" to be the same as ldp:inbox.

Does that clarify things?

**Comment by cwebber** [comment]
> This second bit is about whether an RDF-aware LDN client would be able to discover an LDN inbox for all valid AP 'user profile's.

Does this mean "a type of RDF serialization that is not a json-ld document"?

**Comment by cwebber** [comment]
> So we're clear, [Is this what you intend to do?](https://github.com/w3c/activitystreams/compare/master...gobengo:patch-1)

Btw, that's what @rhiaro has told me in person, that we'd alias as2:inbox to ldp:inbox.  So I think yes, unless I misinterpreted what I heard.

**Comment by rhiaro** [comment]
> So we're clear, [Is this what you intend to do](https://github.com/w3c/activitystreams/compare/master...gobengo:patch-1)?

That is correct.

Adding AS2 extension terms (of which AP is nominally a test case) to the AS2 namespace is a deliberate decision, not a thoughtless one. It is intended to make developers lives easier, when they just want to start using an 'official' extension term, rather than having to introduce a new namespace to their implementation for every extension. Specifications and namespaces are different things, and after a loooong discussion across various lists inside and outside of W3C, in https://github.com/w3c/ldn/issues/13, and in a special session f2f at TPAC in September (specifically about adding `inbox` from LDN to the LDP ns, but also as a general policy) there was consensus that the pros of keeping related terms in the same namespace outweigh the cons. 

@sandhawke and I are working on wording to explain the AS2 extension mechanism - the process for documenting new terms and having the included in the AS2 namespace after the group closes.

> how an AP implementation (that doesn't grok RDF) would try to find an LDN inbox

An implementation that's looking for the plain JSON key `inbox` will only find it if the publisher uses compact JSON-LD to advertise the inbox. We can stress in SWP and/or a note in LDN that if publishers want to receive notifications from naive AP implementations, this is what they need to do.

> whether an RDF-aware LDN client would be able to discover an LDN inbox for all valid AP 'user profile's

If the profile has the AS2 `@context`, once we've added `ldp:inbox` to it, this will be fine.

**Comment by cwebber** [comment]
Ok, since @gobengo claimed that with @rhiaro explaining the position on as:issue and ldp:inbox relationship / extending as2's namespace we could close it, and we got that, so I'm closing this.  Thanks!

**Comment by gobengo** [comment]
> after a loooong discussion across various lists inside and outside of W3C, in w3c/ldn#13, and in a special session f2f at TPAC in September (specifically about adding inbox from LDN to the LDP ns, but also as a general policy) there was consensus that the pros of keeping related terms in the same namespace outweigh the cons.

So it shall be. Thanks for bringing me up to speed :)

I'm satisfied with the closed status. Thanks for your attention @cwebber @rhiaro

@rhiaro wrt adding 'inbox' to AS2 context... it's getting pretty late in the game, right? I don't see an open issue or PR in https://github.com/w3c/activitystreams/issues . Do you want me to make a PR for the branch I linked to, or a placeholder issue?

**Comment by cwebber** [comment]
@gobengo My impression from the namespace conversation at TPAC was that namespaces and vocabularies would become append-only datastructures more or less, and could be extended through some mechanism involving consulting the group who has stewardship over the namespace.  A namespace isn't tied to one standard necessarily, can be extended by multiple forthcoming standards.

There's a lot more to it than that, but that's getting very off-topic for ActivityPub specifically... [the issue Amy linked to](https://github.com/w3c/ldn/issues/13) says more and I'm sure there's minutes for the TPAC meeting out there somewhere...!

**Comment by rhiaro** [comment]
Thanks @bengo... we have an issue for adding all of the AP terms
https://github.com/w3c/activitypub/issues/132
